### PR TITLE
New contexts for `StoreNoticesContainer` and notice grouping

### LINKED ISSUE #2559: All Products: scroll to top when a notice appears
**To Reproduce**
1. In mobile, add a product over its stock quantity with the _All Products_ block.
2. The notice will appear but as a user you can't see it because it's over the fold.

**Expected behavior**
Ideally, the viewport should scroll to the top so the notice is visible to the user.

### PR DESCRIPTION
**Pull request by mikejolley**
This PR refactors `StoreNoticeContainer` so it combines snack bars and regular notices (instead of separate components) and introduces new `noticeContexts` for inner blocks. 

**New noticeContexts**

Adds several notice contexts for the main inner blocks that can be used as `context` instead of strings. Each notice context has a corresponding `StoreNoticesContainer` component to render any notices added.

Rather than create a new interface, I extended the existing one which only contained payment contexts:

```js
PAYMENTS = 'wc/checkout/payments',
EXPRESS_PAYMENTS = 'wc/checkout/express-payments',
```

The full list is now:

```js
CART = 'wc/cart',
CHECKOUT = 'wc/checkout',
PAYMENTS = 'wc/checkout/payments',
EXPRESS_PAYMENTS = 'wc/checkout/express-payments',
CONTACT_INFORMATION = 'wc/checkout/contact-information',
SHIPPING_ADDRESS = 'wc/checkout/shipping-address',
BILLING_ADDRESS = 'wc/checkout/billing-address',
SHIPPING_METHODS = 'wc/checkout/shipping-methods',
CHECKOUT_ACTIONS = 'wc/checkout/checkout-actions',
```

**Notice grouping**

Groups notices of the same type into a single error notice. You can test this on the cart page by calling this a few times:

```
wp.data.dispatch( 'core/notices' ).createNotice( 'error', 'Test', { context: 'wc/cart' } );
```

Screenshot 2022-12-14 at 16 44 13

**Notice creation helpers**

[Existing notices have been modified](https://github.com/woocommerce/woocommerce-blocks/pull/7728) to use the new containers and `createNotice` functions.

Additionally, when a new error is displayed, it will scroll to the parent `StoreNoticeContainer`. The scrolling fixes #2559

This mechanism also ensures that if a `StoreNoticesContainer` does not exist, it uses a parent context.

**Changes to hooks**

**New filters: `showApplyCouponNotice` and `showRemoveCouponNotice`**

To simplify notice handling, I've moved the error suppression for coupons to `showApplyCouponNotice` to prevent the notice from being added in the first place. This will break Points and Rewards (in a non-destructive way), so this will need a patch to use the new filter (https://github.com/woocommerce/woocommerce-points-and-rewards/blob/5d3b6812188e969505f612868ce228a387f1ea12/assets/js/filters.js#L24).

The fix for points and rewards can be found here: https://github.com/woocommerce/woocommerce-points-and-rewards/pull/491

**Other changes**

- Removed `useCheckoutNotices`. It was only used in one place and not shared, so I moved the logic inline.
- `removeAllNotices` when processing checkout so new notices can surface.
- Typed checkout events and push-changes code.
- Helpers to `processErrorResponse` and `processInvalidParamResponse` from the API

Closes #2559

### User-facing testing

Snack bars:

1. Add a product to the cart.
2. Go to cart page
3. Apply a valid coupon code
4. See snack bar notice on the page

Error notices from API:

1. Add a product to the cart.
5. Go to the cart page and set the quantity to 5 for the product.
6. Checkout. Fill out the form but do not submit.
7. In another tab, go into product admin and edit the product. Enable stock management, and set stock qty to 4.
8. Go back to the checkout and place order.
9. Checkout should scroll to the top and display an error:

Screenshot 2022-12-14 at 16 22 28

* [ ] Do not include in the Testing Notes 

cc @opr

### CODE REVIEW

File: packages/checkout/components/store-notices-container/snackbar-notices.tsx
Comment (opr) [inline packages/checkout/components/store-notices-container/snackbar-notices.tsx]: JS linting error here: `[@typescript-eslint/no-unused-vars] 'context' is defined but never used.`

File: packages/checkout/components/store-notices-container/types.ts
Comment (opr) [inline packages/checkout/components/store-notices-container/types.ts]: Can we add a comment to document this please? I understand it's being used to force the notices shown in this component to be a certain type (so if something _was_ a regular notice, but `forceType` is `snackbar` it'll instead turn into a snackbar notice, is this correct?

File: packages/checkout/components/store-notices-container/types.ts
Comment (opr) [inline packages/checkout/components/store-notices-container/types.ts]: If this is an optional property, you should also specify `| undefined` as its type because of this rule https://github.com/woocommerce/woocommerce-blocks/blob/d9f0aa8b8e26bbf50f109ac1f845b22ac8be208b/tsconfig.base.json#L10

same for the other optional props

File: packages/checkout/utils/create-notice.ts
Comment (opr) [inline packages/checkout/utils/create-notice.ts]: What do you think about abstracting this to a function to avoid these if elses?

File: packages/checkout/components/store-notices-container/store-notices.tsx
Comment (opr) [inline packages/checkout/components/store-notices-container/store-notices.tsx]: how come you've named this function? for readability or is there another reason?

File: packages/checkout/components/store-notices-container/snackbar-notices.tsx
Comment (opr) [inline packages/checkout/components/store-notices-container/snackbar-notices.tsx]: The same function is executed with the same args regardless of the condition? Is this is a mistake or am I missing something?

File: assets/js/data/store-notices/reducers.ts
Comment (opr) [inline assets/js/data/store-notices/reducers.ts]: Would it be better to add these constants to a `action-types.ts` file? That's the pattern we used in other data stores.

File: assets/js/data/store-notices/index.ts
Comment (opr) [inline assets/js/data/store-notices/index.ts]: What are these for and why do we need them?

File: assets/js/data/store-notices/default-state.ts
Comment (opr) [inline assets/js/data/store-notices/default-state.ts]: I don't think we should be storing the ref in here, just the context name, everything in a data store should be serializable.

Also: 
image
 :D

File: assets/js/data/store-notices/actions.ts
Comment (opr) [inline assets/js/data/store-notices/actions.ts]: Storing the actual ref object is not right, the data should be serializable. Can we just use the name? I think the result is the same since we only use the object's keys here https://github.com/woocommerce/woocommerce-blocks/blob/4ea76e6a3b7d4254384e8da8252df32bf4dee55d/packages/checkout/utils/create-notice.ts#L32

File: assets/js/base/context/event-emit/utils.ts
Comment (opr) [inline assets/js/base/context/event-emit/utils.ts]: nitpick: should we use the full block name (minus block) for the context? so this would be 

```suggestion
	CONTACT_INFORMATION = 'wc/checkout/contact-information',
```

File: assets/js/base/context/hooks/cart/use-store-cart-coupons.ts
Comment (opr) [inline assets/js/base/context/hooks/cart/use-store-cart-coupons.ts]: Nice, we'll need to sync this with Points and Rewards.

Should we update this deprecation notice too https://github.com/woocommerce/woocommerce-blocks/blob/trunk/packages/checkout/filter-registry/index.ts#L49 in this PR? 

Do you think we should allow `snackbarNoticeVisibility` to also be used here, but with a deprecated notice? I think it's important to deprecate functionality properly, but happy to hear your thoughts on that.

File: assets/js/blocks/checkout/inner-blocks/checkout-contact-information-block/block.tsx
Comment (opr) [inline assets/js/blocks/checkout/inner-blocks/checkout-contact-information-block/block.tsx]: Please see my comment on `assets/js/base/context/event-emit/utils.ts` about possibly renaming this.

**Review by opr — CHANGES_REQUESTED** [review]
Nice Mike, this tested well, but I think we do need to address the refs being in the data store.

We should also deprecate the `snackbarNoticeVisibility` filter properly imo.

Please see my comments below and let me know if anything is unclear or needs further testing.

File: packages/checkout/components/store-notices-container/store-notices.tsx
Comment (mikejolley) [reply packages/checkout/components/store-notices-container/store-notices.tsx]: Yeah I just did it based on an example I found. I'll remove it.

File: assets/js/data/store-notices/index.ts
Comment (mikejolley) [reply assets/js/data/store-notices/index.ts]: I don't actually know. This was in the other stores.

File: assets/js/base/context/hooks/cart/use-store-cart-coupons.ts
Comment (mikejolley) [reply assets/js/base/context/hooks/cart/use-store-cart-coupons.ts]: I think that given the `__experimental` prefix we could replace this without warning. Correct me if I am wrong, but the only side effect to this filter being removed will be a notice displaying right? In which case it's not breaking, just an annoyance. We'll create a patch in points and rewards before this goes live.

Comment (opr) [reply assets/js/base/context/hooks/cart/use-store-cart-coupons.ts]: Ok that sounds reasonable, the extension also uses `__experimentalRegisterCheckoutFilters` to add them so you're right.

**Comment by mikejolley** [comment]
@opr I've worked on everything you mentioned. I'll remove the test buttons before merge, and log/patch points and rewards for the new action name.

**Review by opr — APPROVED** [review]
Nice work, thanks Mike.

I tested it using MMQ and modifying the stock status of various items in my cart.

I am wondering if we should add some testing instructions for Globalstep/release lead for when this gets merged?

**Comment by mikejolley** [comment]
@opr Compatibility patch: https://github.com/woocommerce/woocommerce-points-and-rewards/pull/491

There was a `useMemo` in our filter function which did not work when using filters inline (invalid hook usage). I've removed it — it looks like there is a cached value already so I'm not sure why useMemo was there anyway. Can you check?

**Comment by mikejolley** [comment]
> I tested it using MMQ and modifying the stock status of various items in my cart.

@opr By modifying cart items quantities? We can also trigger a snackbar with an invalid postcode, but we should be fixing that soon I think.

**Comment by opr** [comment]
> @opr By modifying cart items quantities? We can also trigger a snackbar with an invalid postcode, but we should be fixing that soon I think.

Yep, while on the Checkout page, I set the min quantity of an item then tried checking out. It shows the snackbar notice.
image

Setting an incorrect postcode also shows a snackbar.

File: packages/checkout/filter-registry/index.ts
Comment (opr) [inline packages/checkout/filter-registry/index.ts:214]: I just tested this vs. trunk with some filters and they get called the same number of times. I think this is fine to remove.

**Comment by mikejolley** [comment]
@opr I've merged the other notice PR here now - there are 2 tasks blocking this from merging because I don't want to surface too many inline notices as you populate the checkout. The worse ones are email and postcode.

**Review by opr — COMMENTED** [review]
@mikejolley sorry to be pedantic but could we rename the directory **or** store key for the new notices store? Currently the directory names map to  the store name, prefixed with `wc/store`. Based on this folder name I, the data store should be `wc/store/store-notices`. So let's either update the key or dir name.

File: assets/js/types/type-defs/api-response.ts
Comment (opr) [inline assets/js/types/type-defs/api-response.ts:25]: This fails when submitting the checkout form, causing the order to be placed correctly on the server, but the client shows an error.

**Review by opr — CHANGES_REQUESTED** [review]

File: assets/js/types/type-defs/api-response.ts
Comment (opr) [reply assets/js/types/type-defs/api-response.ts:25]: A checkout response looks like this
image

Comment (mikejolley) [reply assets/js/types/type-defs/api-response.ts:25]: Fixed

**Comment by mikejolley** [comment]
@opr This is rebased now, and I've added a description of all the changes I've introduced. There are quite a few files touched, but the changes are mostly minimal. Mostly typescript.

Main changes are new contexts, the merging of notice containers, and the grouping of notices by type.

File: assets/js/base/context/hooks/cart/use-store-cart-coupons.ts
Comment (opr) [inline assets/js/base/context/hooks/cart/use-store-cart-coupons.ts:23]: Do you think it's worth keeping the wording here, just getting rid of the type?

File: docs/third-party-developers/extensibility/checkout-block/available-filters.md
Comment (opr) [inline docs/third-party-developers/extensibility/checkout-block/available-filters.md:215]: Could we document the `showApplyCouponNotice` and `showRemoveCouponNotice` filters in this PR?

**Review by opr — COMMENTED** [review]
Tested this and it worked according to testing instructions.

I think we'll need some design guidance because the notice looks a little weird with the icon and then the list icons, but that can be addressed in another PR, functionally this is OK.

image

Added a couple of inline comments that are non-blocking, and maybe the documentation for the filter could be added in another PR. Approving.

**Review by opr — APPROVED** [review]
Didn't hit approve in the last PR when posting the review 😖

File: assets/js/base/context/hooks/cart/use-store-cart-coupons.ts
Comment (mikejolley) [reply assets/js/base/context/hooks/cart/use-store-cart-coupons.ts:23]: I've moved it inline with the `StoreCartCoupon` type definition.

**Comment by mikejolley** [comment]
@opr that list styling shouldn't be there, so Ill remove that at least and then await design input for other styling changes.

File: docs/third-party-developers/extensibility/checkout-block/available-filters.md
Comment (mikejolley) [reply docs/third-party-developers/extensibility/checkout-block/available-filters.md:215]: Added!

**Pull request merged**
